app/src/main/java/com/rkm/tasky/network/authentication/abstraction/LoginManager.kt

```diff
@@ -0,0 +1,8 @@
+package com.rkm.tasky.network.authentication.abstraction
+
+interface LoginManager {
+    suspend fun logIn(email: String, password: String): Result<Boolean>
+    suspend fun logOut(): Result<Boolean>
```

Comment:
With functions like these, I rarely find it useful to return a boolean result. Its enough to return an empty result and when it does failed, the error returned would indicate that the login failed and also can provide context why it failed. Something like this:

```
suspend fun login(
  email: String,
  password: String,
) : EmptyResult<DataError>
```

You can use it like this:

```
when (val result = api.login(emal, password)) {
  is Result.Success -> {
    // handle login success case
  }

  is Result.Error -> {
    // handle login failed case
  }
}
```

This result class is based on a generic error handling wrapper from Philipp which I’ve found more useful than Kotlin’s `Result` class.

```
sealed interface Result<out D, out E: Error> {
    data class Success<out D>(val data: D): Result<D, Nothing>
    data class Error<out E: com.plcoding.tasky.core.domain.util.Error>(val error: E): Result<Nothing, E>
}

inline fun <T, E: Error, R> Result<T, E>.map(map: (T) -> R): Result<R, E> {
    return when(this) {
        is Result.Error -> Result.Error(error)
        is Result.Success -> Result.Success(map(data))
    }
}

fun <T, E: Error> Result<T, E>.asEmptyDataResult(): EmptyResult<E> {
    return map {  }
}

inline fun <T, E: Error> Result<T, E>.onSuccess(action: (T) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}
inline fun <T, E: Error> Result<T, E>.onError(action: (E) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}

typealias EmptyResult<E> = Result<Unit, E>
```
It handles data and errors in a type-safe way. You'll need a marker interface to mark error enums:
```
interface Error
```
And then you can extend this across the codebase with different error types, e.g.:
```
sealed interface DataError: Error {
    enum class Remote: DataError {
        NOT_FOUND,
        NO_INTERNET,
        UNAUTHORIZED,
        ...
    }
    
    enum class Local: DataError {
        DISK_FULL,
        ...
    }
}
```
Works great:
```
repository.signUp(payload).onSuccess { responseData ->

}.onError { error -> 
    // toUiText() can be an extension of DataError which maps an error to the corresponding message
    _state.update { it.copy(error = error.toUiText()) }
}
```